instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/InstrumenterBuilder.java

```diff
@@ -190,39 +190,57 @@ public InstrumenterBuilder<REQUEST, RESPONSE> setEnabled(boolean enabled) {
    * and inject context into requests.
    */
   public Instrumenter<REQUEST, RESPONSE> buildClientInstrumenter(TextMapSetter<REQUEST> setter) {
-    return buildInstrumenter(
-        InstrumenterConstructor.propagatingToDownstream(requireNonNull(setter, "setter")),
-        SpanKindExtractor.alwaysClient());
+    return buildOngoingInstrumenter(
+        requireNonNull(setter, "setter"), SpanKindExtractor.alwaysClient());
   }
 
   /**
    * Returns a new {@link Instrumenter} which will create {@linkplain SpanKind#SERVER server} spans
    * and extract context from requests.
    */
   public Instrumenter<REQUEST, RESPONSE> buildServerInstrumenter(TextMapGetter<REQUEST> getter) {
-    return buildInstrumenter(
-        InstrumenterConstructor.propagatingFromUpstream(requireNonNull(getter, "getter")),
-        SpanKindExtractor.alwaysServer());
+    return buildIncomingInstrumenter(
+        requireNonNull(getter, "getter"), SpanKindExtractor.alwaysServer());
   }
 
   /**
    * Returns a new {@link Instrumenter} which will create {@linkplain SpanKind#PRODUCER producer}
    * spans and inject context into requests.
    */
   public Instrumenter<REQUEST, RESPONSE> buildProducerInstrumenter(TextMapSetter<REQUEST> setter) {
-    return buildInstrumenter(
-        InstrumenterConstructor.propagatingToDownstream(requireNonNull(setter, "setter")),
-        SpanKindExtractor.alwaysProducer());
+    return buildOngoingInstrumenter(
+        requireNonNull(setter, "setter"), SpanKindExtractor.alwaysProducer());
   }
 
   /**
    * Returns a new {@link Instrumenter} which will create {@linkplain SpanKind#CONSUMER consumer}
    * spans and extract context from requests.
    */
   public Instrumenter<REQUEST, RESPONSE> buildConsumerInstrumenter(TextMapGetter<REQUEST> getter) {
+    return buildIncomingInstrumenter(
+        requireNonNull(getter, "getter"), SpanKindExtractor.alwaysConsumer());
+  }
+
+  /**
+   * Returns a new {@link Instrumenter} which will create spans with kind determined by the passed
+   * {@link SpanKindExtractor} and inject context into requests.
+   */
+  public Instrumenter<REQUEST, RESPONSE> buildOngoingInstrumenter(
+      TextMapSetter<REQUEST> setter, SpanKindExtractor<? super REQUEST> spanKindExtractor) {
+    return buildInstrumenter(
+        InstrumenterConstructor.propagatingToDownstream(requireNonNull(setter, "setter")),
+        requireNonNull(spanKindExtractor, "spanKindExtractor"));
+  }
+
+  /**
+   * Returns a new {@link Instrumenter} which will create spans with kind determined by the passed
+   * {@link SpanKindExtractor} and extract context from requests.
+   */
+  public Instrumenter<REQUEST, RESPONSE> buildIncomingInstrumenter(
```

Comment:
I am not sure if it is ok to add such a public method to the InstrumenterBuilder, but I could not think of a better way to make SpanKindExtractor configurable.

Reply:
Yeah, it is necessary to add a new method here (I kinda forgot about it earlier, when we were talking in GH discussions).

Since this is a stable public API, we need to make sure we come up with a good name. I'll list a few ideas I came up with here:
* `buildIncomingInstrumenter`/`buildOutgoingInstrumenter` (pretty much what this PR proposes)
* `buildContextExtractingInstrumenter`/`buildContextInjectingInstrumenter`
* `buildUpstreamInstrumenter`/`buildDownstreamInstrumeter`
* `buildPropagatingToUpstreamInstrumenter`/`buildPropagatingToDownstreamInstrumeter`
* `buildPropagatingInstrumenter` for both

@open-telemetry/java-instrumentation-maintainers WDYT?

Reply:
Naming is hard. Perhaps we should adopt a `@Incubating` or `@Experimental` annotations so such changes could be introduced more easily.
I think it is possible to implement this feature without introducing new api. You could use `buildInstrumenter(spanKindExtractor)` and do the context extraction manually before `instrumenter.start()` similarly to like it is done in `PropagatingFromUpstreamInstrumenter`.

Reply:
@laurit  Thank you for your suggestion. I tried implementing this without introducing a new API.
https://github.com/orangain/opentelemetry-java-instrumentation/commit/daec120a46dca5c5581b0ed9880dec3e81ee65b4

While it is indeed possible to implement, I have a few concerns.

* Doing context propagation within library instrumentation seems to be a violation of responsibilities. Wouldn't it be confusing if only Ktor instrumentation does context propagation on its own, while `Instrumenter` does it in other library instrumentations?
* `ContextPropagationDebug` exists in the `internal` package, is it acceptable to use it in Ktor instrumentation? If not allowed, there is a concern that Ktor users will not be able to debug context propagation.

Reply:
@orangain While letting `Instrumenter` handle context propagation is preferable there are instrumentations that do it manually for a variety of reasons. This was just a suggestion on how to move this forward faster, if you so wish you can ignore it. You could also start with manual propagation and when the methods you need are added to `Instrumenter` you can start using them.
Using internal classes is fine everywhere in this repository. Marking classes internal is more for informing users outside this repository that these classes could change without prior notice. I think that in this case not using `ContextPropagationDebug` would also be fine.

Reply:
Hey @orangain I opened #8392 to unblock you PR -- I'll let you know when it's merged so that you can rebase on top of it.

Reply:
hey @orangain, #8392 is merged now

Reply:
@mateuszrzeszutek @trask @laurit  Sorry for the late response. Thank you for unblocking my PR! I've rebased the PR to use `InstrumenterUtil.buildUpstreamInstrumenter`.